app/celery/nightly_tasks.py

```diff
@@ -334,3 +336,63 @@ def save_daily_notification_processing_time(bst_date=None):
             messages_within_10_secs=result.messages_within_10_secs,
         )
     )
+
+
+@notify_celery.task(name="delete-oldest-quarter-of-unneeded-notification-history")
+def delete_oldest_quarter_of_unneeded_notification_history():
+    # This function will delete either 1 or 0 quarters of notifications from the notification_history
+    # table. It will pick the oldest quarter available before our retention limit and delete that.
+    # If there are no notifications older than our retention limit, then it won't do anything.
+    # By running this task several nights in a row, it will slowly delete a quarter of notification_history
+    # per night until we reach out retention limit
+    #
+    # We do everything in this function in BST up until the final point
+    # This is because we care about quarters in the BST timezone because this is
+    # how we bill our users and is easier to reason about.
+    # When we finally figure out which BST quarter we can delete, we convert the time
+    # to UTC so that we delete the correct notifications from our database
+    #
+    # This retention limit is hardcoded and was originally picked from
+    # https://github.com/alphagov/notifications-aws/blob/main/decisions/2022-12-01-notification-history-retention-period.md
+    # It was supposed to be 2023-4-1.
+    # However, at the time of writing this code we realised the FtBillingLetterDispatch table has been introduced
+    # which means we need to store letters older than 2023-4-1 in order to rebuild that table (because that table
+    # uses the date of dispatch, not the date of creation for which date to bill for). To keep it simple, we keep
+    # an extra quarters worth of data giving us plenty of buffer
+    #
+    # In the future, we will be able to update this retention_limit_bst value when we have progressed 3 quarters
+    # into the next financial year
+    retention_limit_bst = datetime(2023, 1, 1)
+
+    oldest_notification_bst = convert_utc_to_bst(db.session.query(func.min(NotificationHistory.created_at)).one()[0])
```

Comment:
Maybe we could add log line that will tell us when the oldest notification was from? That could be useful for debugging. For example:

`delete_oldest_quarter_of_unneeded_notification_history: Oldest notification found in notification_history table was created_at {oldest_notification_bst} BST.`

Reply:
Possibly logging the time in UTC would be more useful, since that's what we would see in the DB and generally what we deal with behind the scenes, I think?

Reply:
I agree 👍🏼

Reply:
Will try and follow up with a subsequent PR to add in that logging...

I agree it is useful but this PR has taken me soo long to find time to merge that I'm going to merge as is (otherwise it won't happen for another few days)...